backend/sessions.md

```diff
@@ -0,0 +1,129 @@
+---
+layout: page
+title: Tips
+nav_order: 550
+parent: Sessions
+---
+
+# Sessions
+
+The service that creates and refresh the session is the same: `accessTokenCreate`.
+
+A user can have multiple sessions, typically one per device.
+
+A session can have multiple access tokens. The access tokens have a lifetime of 2 hours,
+after which they have to be refreshed. There can be more than one valid access token for a session at the same time,
+but we always want to use the most recent one.
+
+If there's no session, we create a temporary user, and a session is created for it.
+
+If a `code` is sent (coming from the login-module by the frontend after authentication), the session is created for the user that logged in,
+as described in the section [Creation of a session for a normal user (not a temporary user)](#creation-of-a-session-for-a-normal-user-not-a-temporary-user).
+
+If there's already a session for the user (and no `code` is sent), the process is described in the section [Refreshing the token](#refreshing-the-token).
+
+
+## Creation of a session for a normal user (not a temporary user)
+
+A session is created when a user logs in.
+
+This is the process of creating a session:
+
+1. We call the login-module to get back:
+  - the access token
+  - the refresh token
+  - the expiration time of the access token
+2. If the user doesn't exist in the database, we create it, otherwise we update the `latest_login_at`, `latest_activity_at`, and `latest_profile_sync_at` fields.
+3. The session is created in the database with the refresh token. The access token is added to the session.
+4. If there are already 10 sessions for the user, or more, we delete the oldest ones, to have a maximum of 10 sessions in database.
+5. We return the access token with its expiration time.
+
+
+### There's a limit of 10 concurrent sessions per user
+
+We limit the number of concurrent sessions per user to 10.
+
+This is to avoid bloating the database with too many sessions.
+
+
+## Refreshing the token
+
+This is the process of refreshing the token:
+
+1. (Authentication middleware) We check that the token is valid and not expired (same as any authenticated request).
+2. We retrieve the most recent token of the session. If the token used to call the refresh is not the most recent one,
+   we return the most recent token with its expiration time.
+3. (From now on, we're working with the most recent token). If the token has been issued less than 5 minutes ago,
+   we return the token with its expiration time.
+4. We get a lock for the session.
+5. We call the login-module to refresh the token, with the refresh token of the session, and get back:
+  - the new access token
+  - the new refresh token
+  - the expiration time of the access token
+6. We store the refresh token in the session, and store the new access token.
+7. We delete expired old tokens for the session.
```

Comment:
You get a lock but you don't say when you release it.
Getting a lock for a remote access is dangerous (if you loose request, you process is killed, ...), how do you protect against that?